contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+                .fold(
+                    Coin {
+                        amount: Uint128(0),
+                        denom: DENOM.to_string(),
+                    },
+                    |mut acc, d| {
+                        acc.amount += d.amount;
+                        acc
+                    },
+                )
+        } else {
+            Coin {
+                amount: Uint128(0),
+                denom: DENOM.to_string(),
+            }
+        }
+    }
+
+    fn get_balance(&self, storage: &dyn Storage) -> Uint128 {
+        get_account_balance(storage, &self.address).unwrap_or(Uint128(0))
+    }
+}
+
+impl DelegationAccount for PeriodicVestingAccount {
+    fn try_delegate_to_mixnode(
+        &self,
+        mix_identity: IdentityKey,
+        coin: Coin,
+        env: &Env,
+        storage: &mut dyn Storage,
+        querier: Option<QuerierWrapper>,
+    ) -> Result<(), ContractError> {
+        if coin.amount < self.get_balance(storage) {
+            if let Some(querier) = querier {
+                let msg = MixnetExecuteMsg::DelegateToMixnodeOnBehalf {
+                    mix_identity: mix_identity.clone(),
+                    delegate_addr: self.address.clone(),
+                    coin: coin.clone(),
+                };
+                querier.query_wasm_smart(DEFAULT_MIXNET_CONTRACT_ADDRESS, &msg)?;
+            }
+            self.track_delegation(env.block.time, mix_identity, coin, storage)?;
+            Ok(())
+        } else {
+            return Err(ContractError::InsufficientBalance(
+                self.address.as_str().to_string(),
+                self.get_balance(storage).u128(),
+            ));
+        }
+    }
+
+    fn try_undelegate_from_mixnode(
+        &self,
+        mix_identity: IdentityKey,
+        storage: &mut dyn Storage,
+        querier: Option<QuerierWrapper>,
+    ) -> Result<(), ContractError> {
+        if let Some(querier) = querier {
+            let msg = MixnetExecuteMsg::UnDelegateFromMixnodeOnBehalf {
+                mix_identity: mix_identity.clone(),
+                delegate_addr: self.address.clone(),
+            };
+            querier.query_wasm_smart(DEFAULT_MIXNET_CONTRACT_ADDRESS, &msg)?;
+        }
+        self.track_undelegation(mix_identity, storage)?;
+        Ok(())
+    }
+
+    fn track_delegation(
+        &self,
+        block_time: Timestamp,
+        mix_identity: IdentityKey,
+        delegation_amount: Coin,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        let mut delegations =
+            if let Some(delegations) = get_account_delegations(storage, &self.address) {
+                delegations
+            } else {
+                Vec::new()
+            };
+        delegations.push(DelegationData {
+            mix_identity,
+            amount: delegation_amount.amount,
+            block_time,
+        });
+        set_account_delegations(storage, &self.address, delegations)?;
```

Comment:
This looks a bit suboptimal, to re-write the entire vector every time. I think this could be done better with a multilevel bucket, otherwise costs would go up constantly for this function.

Reply:
Yeah, it very much is, I just really don't like multilevel buckets, but you're right, this has the potential to really slow us down

Reply:
@neacsu, @jstuczyn I've been noodling on this some more, and the most common case here is iterating through all delegations to find which are vesting/vested, for that one vector is the ideal data structure. As far as a performance hit, even hundreds of delegations should not take a lot of storage space, so read/write ops should be fairly fast. I'd like to benchmark this somehow, maybe implement pages as vectors instead a multilevel bucket, which would be pretty nasty to work with, and would bring along at lest N reads for every time we want to figure out vested/vesting tokens, as well as any derived function, such as `delegated_vesting/delegated_free/locked/spendable`. I'll bring this to cosmwasm folks before doing any changes, as I'd really like to understand the constraints involved... Dissenting opinions more the welcome of coourse :)